src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -385,91 +385,65 @@ Node TheoryStringsRewriter::rewriteStrEqualityExt(Node node)
   // ------- homogeneous constants
   for (unsigned i = 0; i < 2; i++)
   {
-    if (node[i].isConst())
+    Node cn = checkEntailHomogeneousString(node[i]);
+    if (!cn.isNull() && cn.getConst<String>().size() > 0)
     {
-      bool isHomogeneous = true;
-      unsigned hchar = 0;
-      String lhss = node[i].getConst<String>();
-      std::vector<unsigned> vec = lhss.getVec();
-      if (vec.size() >= 1)
-      {
-        hchar = vec[0];
-        for (unsigned j = 1, size = vec.size(); j < size; j++)
-        {
-          if (vec[j] != hchar)
-          {
-            isHomogeneous = false;
-            break;
-          }
-        }
-      }
-      if (isHomogeneous)
+      Assert(cn.isConst());
+      Assert(cn.getConst<String>().size() == 1);
+      unsigned hchar = cn.getConst<String>().front();
+
+      std::vector<Node> trimmed[2];
+      size_t numHChars[2] = {0, 0};
+      for (size_t j = 0; j < 2; j++)
```

Comment:
Can you comment a bit on what this loop is doing?  Took me a while to figure this out in the new version.

Reply:
Done